### ISSUE TITLE
Unexpected pageContext behaviour when navigating to page via the URL bar and back

### ISSUE THREAD

**Issue by jgalminas** [vikejs/vike#1275]
### Description

I ran into this issue when implementing authentication following the [auth example](https://github.com/vikejs/vike/tree/main/examples/auth).

After logging in the user object is defined in pageContext and the Vike reload function is called. From there, if the user navigates to any page by typing in the URL in the address bar and pressing enter and then navigates back, the user object which was set in pageContext becomes null.

I'm using the React preset and V1 just like in the auth example. In fact, the same issue can be reproduced in the example by doing the following:

1. Load up the app and navigate to the log in page `/login` via the URL bar.
2. Log in (doesn't matter which one).
3. Navigate to `/` via the URL bar again.
4. Navigate back.

Thanks :)

**Comment by AurelienLourot** [comment]
Thanks, I can reproduce and we'll have a look tomorrow :)

**Comment by brillout** [comment]
What's your Vike version? See https://twitter.com/brillout/status/1724469728603754550.

**Comment by AurelienLourot** [comment]
@brillout I can reproduce this on `main`, but really exactly with these steps. Any other combinations of backward-forward is fine

**Comment by brillout** [comment]
Ok it's unrelated then. Let me know if you want me to have a look at it myself.

**Comment by AurelienLourot** [comment]
Actually #1267 fixes this issue as a side-effect, thanks to the changes made there to `getPageContextFromHooks.ts`. Do you confirm @jgalminas that if you check out the branch `aurelien/feature/data-hook` and run the `auth` example there the problem is gone?

**Comment by jgalminas** [comment]
@AurelienLourot Hey. I can still reproduce it.

**Comment by AurelienLourot** [comment]
You're right, not sure what I did wrong. #1267 doesn't fix the problem. I'll have another look

**Comment by AurelienLourot** [comment]
@brillout quick check if this rings a bell. It doesn't happen when I have the developer tools open, because then some caching is disabled, and it always happens when the developer tools are closed. Here is in details what happens:

1. start the example with `pnpm run dev`

2. open manually http://localhost:3000/login

The server delivers a pageContext with `userFullName undefined`. The client runs `getPageContextFromHooks_firstRender()`. We are logged out, all good ✅

3. Log in as Alan Turing

The client runs `getPageContextFromHooks_uponNavigation()` and fetches `/login/index.pageContext.json` which contains `userFullName Alan Turing`. We are still on page `/login`, we are logged in, all good  ✅

4. **Type manually in the address bar** https://localhost:3000/

The server still delivers a pageContext with `userFullName Alan Turing` thanks to the cookie. The client runs `getPageContextFromHooks_firstRender()`. We are still logged in, all good ✅

5. **Hit the [back] button**

The problem here is that the browser knows that the previous page is `/login` and remembers its HTML from step 2. So it just delivers that HTML that is cached and no communication whatsoever is done with the server. Unfortunately in that HTML the pageContext says `userFullName undefined` ❌ The client runs `getPageContextFromHooks_firstRender()`. We appear mistakenly as logged out ❌

6. Hit [refresh]

Communication is made with the server again which delivers pageContext with `userFullName Alan Turing` again. We appear as logged in again ✅

It seems to be a very edge-case because following conditions must be fulfilled:
* Navigate manually to `/login` while being logged out, so that the browser will cache this HTML (step 2)
* While logged in later, hit [back] when the previous page is `/login` (step 5) and the browser will just deliver the stale HTML with stale pageContext

@brillout Do you have an idea how to mitigate this off the top of your head or a pointer? Otherwise I'll dig myself further. Thanks :pray:

**Comment by AurelienLourot** [comment]
Nevermind @brillout, I think I have figured it out.

> It seems to be a very edge-case because following conditions must be fulfilled: [...]

Since this can only happen on the `/login` page, I think the user (here the `auth` example) should tell Vike to set the HTTP headers on that particular page so that it doesn't get cached. I guess this has to be done in the server glue code https://github.com/vikejs/vike/blob/main/examples/auth/server/index.js#L30 . Let me give it a try.

**Comment by AurelienLourot** [comment]
TL;DR actually it can happen on any page. I'm about to create a PR disabling caching for all pages in the `auth` example.

If you copy the `index` page to a new `hello` page, you can reproduce it as follows:
1. type manually http://localhost:3000/hello while logged out (the page gets cached)
2. type manually http://localhost:3000/login
3. log in (you are still on the /login page)
4. hit [back] -> the browser shows you the logged out version of the /hello page that was cached on step 1 ❌

Now for a more complex app, the developer knows which pages are safe to cache but for this particular `auth` example it seems to me that none of all the pages is safe to cache.

**Comment by brillout** [comment]
@AurelienLourot Very neat digging and explanation btw. it makes perfect sense.
